File: Code/benchmarks/binary_trees/bt_arena.cpp
// Arena C++: bump-allocate nodes from a per-tree arena, bulk-freed (matches Elisa's region model).
#include <cstdio>
#include <cstdlib>
#include <vector>
struct Tree { Tree* l; Tree* r; };
struct Arena {
    std::vector<char*> blocks; size_t off=0, cap=0; char* cur=nullptr;
    static const size_t BLK = 1<<20;
    Tree* alloc(){
        if(off+sizeof(Tree)>cap){ cur=(char*)malloc(BLK); blocks.push_back(cur); cap=BLK; off=0; }
        Tree* t=(Tree*)(cur+off); off+=sizeof(Tree); return t;
    }
    ~Arena(){ for(char* b:blocks) free(b); }
};
static Tree* make(Arena& a, long depth){
    Tree* t=a.alloc();
    if(depth>0){ t->l=make(a,depth-1); t->r=make(a,depth-1); } else { t->l=nullptr; t->r=nullptr; }
    return t;
}
static long check(const Tree* n){ if(!n->l) return 1; return 1+check(n->l)+check(n->r); }
int main(){
    long max_depth=18, min_depth=4, total=0;
    { Arena a; auto s=make(a,max_depth+1); total+=check(s); }
    Arena longArena; auto longLived=make(longArena,max_depth);
    for(long d=min_depth; d<=max_depth; d+=2){
        long iters=1L<<(max_depth-d+min_depth), sum=0;
        for(long i=0;i<iters;i++){ Arena a; auto t=make(a,d); sum+=check(t); }
        total+=sum;
    }
    total+=check(longLived);
    printf("binary-trees(18) check total: %ld\n", total);
    return 0;
}

File: Code/benchmarks/binary_trees/bt_arena.rs
// Index-arena Rust: nodes in one Vec, children as indices; arena dropped per tree (bulk free).
struct Node { left: i32, right: i32 } // left < 0 => leaf
fn make(a: &mut Vec<Node>, d: i64) -> i32 {
    if d <= 0 { a.push(Node{left:-1,right:-1}); (a.len()-1) as i32 }
    else { let l = make(a,d-1); let r = make(a,d-1); a.push(Node{left:l,right:r}); (a.len()-1) as i32 }
}
fn check(a: &Vec<Node>, i: i32) -> i64 {
    let n = &a[i as usize];
    if n.left < 0 { 1 } else { 1 + check(a, n.left) + check(a, n.right) }
}
fn main() {
    let (max_depth, min_depth) = (18i64, 4i64);
    let mut total;
    { let mut a = Vec::new(); let s = make(&mut a, max_depth+1); total = check(&a, s); }
    let mut la = Vec::new(); let ll = make(&mut la, max_depth);
    let mut d = min_depth;
    while d <= max_depth {
        let iters = 1i64 << (max_depth - d + min_depth);
        let mut s = 0i64;
        for _ in 0..iters { let mut a = Vec::new(); let t = make(&mut a, d); s += check(&a, t); }
        total += s;
        d += 2;
    }
    total += check(&la, ll);
    println!("binary-trees(18) check total: {}", total);
}
